File: lib/live_set.rb
require_relative 'live_set/live_set_options'
require_relative 'common/util'
require_directory "#{__dir__}/common"

def common(command)
  @options = parse_options command
  help_live_set 'Ableton Live set name must be provided.' if ARGV.empty?
end

def als_delta
  require_directory "#{__dir__}/als_delta"

  help_als_delta if ARGV.empty?

  @options = parse_options

  set_name = ARGV.join(' ').expand_env
  help_als_delta "#{set_name} does not exist" unless File.exist? set_name
  help_als_delta "#{set_name} is a directory" if File.directory? set_name

  trap('SIGINT') { throw :ctrl_c }

  als_delta = AlsDelta.new set_name, **@options
  catch :ctrl_c do
    als_delta.show
    als_delta.cleanup
  rescue Exception # rubocop:disable Lint/RescueException
    als_delta.cleanup
  end
end

def live_set
  require_directory "#{__dir__}/live_set"

  help_live_set if ARGV.empty?

  @options = parse_options

  set_name = ARGV.join(' ').expand_env
  help_live_set "#{set_name} does not exist" unless File.exist? set_name
  help_live_set "#{set_name} is a directory" if File.directory? set_name

  live_set = LiveSet.new set_name, **@options

  if @options[:convert11]
    live_set.modify_als
  else
    live_set.show
  end
end

File: src/live_set.rs
use std::env;
use std::fs;
use std::path::Path;
use std::process;

mod live_set;
mod als_delta;
mod common;

fn main() {
    let command = env::args().nth(1).unwrap_or_else(|| {
        eprintln!("Please provide a command (common, als_delta, or live_set).");
        process::exit(1);
    });

    match command.as_str() {
        "common" => common(),
        "als_delta" => als_delta(),
        "live_set" => live_set(),
        _ => {
            eprintln!("Invalid command. Please use 'common', 'als_delta', or 'live_set'.");
            process::exit(1);
        }
    }
}

fn common() {
    let options = common::parse_options();
    if env::args().len() == 1 {
        common::help_live_set("Ableton Live set name must be provided.");
    }
}

fn als_delta() {
    let options = als_delta::parse_options();
    if env::args().len() == 1 {
        als_delta::help_als_delta();
    }

    let set_name = env::args().skip(1).collect::<Vec<_>>().join(" ").expand_env();
    if !Path::new(&set_name).exists() {
        als_delta::help_als_delta(format!("{} does not exist", set_name));
    }
    if Path::new(&set_name).is_dir() {
        als_delta::help_als_delta(format!("{} is a directory", set_name));
    }

    let mut als_delta = als_delta::AlsDelta::new(&set_name, &options);
    match catch_ctrl_c!(als_delta.show()) {
        Ok(_) => als_delta.cleanup(),
        Err(e) => {
            als_delta.cleanup();
            eprintln!("Error: {}", e);
        }
    }
}

fn live_set() {
    let options = live_set::parse_options();
    if env::args().len() == 1 {
        live_set::help_live_set();
    }

    let set_name = env::args().skip(1).collect::<Vec<_>>().join(" ").expand_env();
    if !Path::new(&set_name).exists() {
        live_set::help_live_set(format!("{} does not exist", set_name));
    }
    if Path::new(&set_name).is_dir() {
        live_set::help_live_set(format!("{} is a directory", set_name));
    }

    let mut live_set = live_set::LiveSet::new(&set_name, &options);
    if options.convert11 {
        live_set.modify_als();
    } else {
        live_set.show();
    }
}

trait ExpandEnv {
    fn expand_env(self) -> String;
}

impl ExpandEnv for String {
    fn expand_env(self) -> String {
        let mut result = self;
        for (key, value) in env::vars() {
            result = result.replace(&format!("${}", key), &value);
        }
        result
    }
}

macro_rules! catch_ctrl_c {
    ($expr:expr) => {
        catch::catch(|| $expr)
    };
}
